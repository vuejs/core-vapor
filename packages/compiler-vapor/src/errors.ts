import {
  CompilerError,
  SourceLocation,
  createCompilerError,
} from '@vue/compiler-dom'

export interface VaporCompilerError extends CompilerError {
  code: VaporErrorCodes
}

export function createVaporCompilerError(
  code: VaporErrorCodes,
  loc?: SourceLocation,
) {
  return createCompilerError(
    code,
    loc,
    __DEV__ || !__BROWSER__ ? VaporErrorMessages : undefined,
  ) as VaporCompilerError
}

export enum VaporErrorCodes {
  X_V_PLACEHOLDER = 100,
  __EXTEND_POINT__,
}

export const VaporErrorMessages: Record<VaporErrorCodes, string> = {
  [VaporErrorCodes.X_V_PLACEHOLDER]: `[placeholder]`,

  // just to fulfill types
  [VaporErrorCodes.__EXTEND_POINT__]: ``,
}
