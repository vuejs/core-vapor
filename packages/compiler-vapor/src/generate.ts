import {
  type CodegenOptions,
  type CodegenResult,
  type Position,
  type SourceLocation,
  NewlineType,
  advancePositionWithMutation,
  locStub,
  NodeTypes,
  BindingTypes,
  isSimpleIdentifier,
} from '@vue/compiler-dom'
import {
  type IRDynamicChildren,
  type RootIRNode,
  IRNodeTypes,
  OperationNode,
  VaporHelper,
  IRExpression,
} from './ir'
import { SourceMapGenerator } from 'source-map-js'
import { isString } from '@vue/shared'

// remove when stable
// @ts-expect-error
function checkNever(x: never): never {}

export interface CodegenContext extends Required<CodegenOptions> {
  source: string
  code: string
  line: number
  column: number
  offset: number
  indentLevel: number
  map?: SourceMapGenerator

  push(
    code: string,
    newlineIndex?: number,
    loc?: SourceLocation,
    name?: string,
  ): void
  pushWithNewline(
    code: string,
    newlineIndex?: number,
    loc?: SourceLocation,
    name?: string,
  ): void
  indent(): void
  deindent(): void
  newline(): void

  helpers: Set<string>
  vaporHelpers: Set<string>
  helper(name: string): string
  vaporHelper(name: string): string
}

function createCodegenContext(
  ir: RootIRNode,
  {
    mode = 'function',
    prefixIdentifiers = mode === 'module',
    sourceMap = false,
    filename = `template.vue.html`,
    scopeId = null,
    optimizeImports = false,
    runtimeGlobalName = `Vue`,
    runtimeModuleName = `vue`,
    ssrRuntimeModuleName = 'vue/server-renderer',
    ssr = false,
    isTS = false,
    inSSR = false,
    inline = false,
    bindingMetadata = {},
  }: CodegenOptions,
) {
  const { helpers, vaporHelpers } = ir
  const context: CodegenContext = {
    mode,
    prefixIdentifiers,
    sourceMap,
    filename,
    scopeId,
    optimizeImports,
    runtimeGlobalName,
    runtimeModuleName,
    ssrRuntimeModuleName,
    ssr,
    isTS,
    inSSR,
    bindingMetadata,
    inline,

    source: ir.source,
    code: ``,
    column: 1,
    line: 1,
    offset: 0,
    indentLevel: 0,

    helpers,
    vaporHelpers,
    helper(name: string) {
      helpers.add(name)
      return `_${name}`
    },
    vaporHelper(name: VaporHelper) {
      vaporHelpers.add(name)
      return `_${name}`
    },
    push(code, newlineIndex = NewlineType.None, loc, name) {
      context.code += code
      if (!__BROWSER__ && context.map) {
        if (loc) addMapping(loc.start, name)

        if (newlineIndex === NewlineType.Unknown) {
          // multiple newlines, full iteration
          advancePositionWithMutation(context, code)
        } else {
          // fast paths
          context.offset += code.length
          if (newlineIndex === NewlineType.None) {
            // no newlines; fast path to avoid newline detection
            if (__TEST__ && code.includes('\n')) {
              throw new Error(
                `CodegenContext.push() called newlineIndex: none, but contains` +
                  `newlines: ${code.replace(/\n/g, '\\n')}`,
              )
            }
            context.column += code.length
          } else {
            // single newline at known index
            if (newlineIndex === NewlineType.End) {
              newlineIndex = code.length - 1
            }
            if (
              __TEST__ &&
              (code.charAt(newlineIndex) !== '\n' ||
                code.slice(0, newlineIndex).includes('\n') ||
                code.slice(newlineIndex + 1).includes('\n'))
            ) {
              throw new Error(
                `CodegenContext.push() called with newlineIndex: ${newlineIndex} ` +
                  `but does not conform: ${code.replace(/\n/g, '\\n')}`,
              )
            }
            context.line++
            context.column = code.length - newlineIndex
          }
        }
        if (loc && loc !== locStub) {
          addMapping(loc.end)
        }
      }
    },
    pushWithNewline(code, newlineIndex, node) {
      context.newline()
      context.push(code, newlineIndex, node)
    },
    indent() {
      ++context.indentLevel
    },
    deindent() {
      --context.indentLevel
    },
    newline() {
      newline(context.indentLevel)
    },
  }

  function newline(n: number) {
    context.push(`\n${`  `.repeat(n)}`, NewlineType.Start)
  }

  function addMapping(loc: Position, name: string | null = null) {
    // we use the private property to directly add the mapping
    // because the addMapping() implementation in source-map-js has a bunch of
    // unnecessary arg and validation checks that are pure overhead in our case.
    const { _names, _mappings } = context.map!
    if (name !== null && !_names.has(name)) _names.add(name)
    _mappings.add({
      originalLine: loc.line,
      originalColumn: loc.column - 1, // source-map column is 0 based
      generatedLine: context.line,
      generatedColumn: context.column - 1,
      source: filename,
      // @ts-ignore it is possible to be null
      name,
    })
  }

  if (!__BROWSER__ && sourceMap) {
    // lazy require source-map implementation, only in non-browser builds
    context.map = new SourceMapGenerator()
    context.map.setSourceContent(filename, context.source)
    context.map._sources.add(filename)
  }

  return context
}

// IR -> JS codegen
export function generate(
  ir: RootIRNode,
  options: CodegenOptions = {},
): CodegenResult {
  const ctx = createCodegenContext(ir, options)
  const { push, pushWithNewline, indent, deindent, newline } = ctx
  const { vaporHelper, helpers, vaporHelpers } = ctx

  const functionName = 'render'
  const isSetupInlined = !!options.inline
  if (isSetupInlined) {
    push(`(() => {`)
  } else {
    // placeholder for preamble
    newline()
    pushWithNewline(`export function ${functionName}(_ctx) {`)
  }
  indent()

  ir.template.forEach((template, i) => {
    if (template.type === IRNodeTypes.TEMPLATE_FACTORY) {
      // TODO source map?
      pushWithNewline(
        `const t${i} = ${vaporHelper('template')}(${JSON.stringify(
          template.template,
        )})`,
      )
    } else {
      // fragment
      pushWithNewline(
        `const t0 = ${vaporHelper('fragment')}()\n`,
        NewlineType.End,
      )
    }
  })

  {
    pushWithNewline(`const n${ir.dynamic.id} = t0()`)

    const children = genChildren(ir.dynamic.children)
    if (children) {
      pushWithNewline(
        `const ${children} = ${vaporHelper('children')}(n${ir.dynamic.id})`,
      )
    }

    for (const operation of ir.operation) {
      genOperation(operation, ctx)
    }
    for (const { operations } of ir.effect) {
      pushWithNewline(`${vaporHelper('effect')}(() => {`)
      indent()
      for (const operation of operations) {
        genOperation(operation, ctx)
      }
      deindent()
      pushWithNewline('})')
    }
    // TODO multiple-template
    // TODO return statement in IR
    pushWithNewline(`return n${ir.dynamic.id}`)
  }

  deindent()
  newline()
  if (isSetupInlined) {
    push('})()')
  } else {
    push('}')
  }

  let preamble = ''
  if (vaporHelpers.size)
    // TODO: extract import codegen
    preamble = `import { ${[...vaporHelpers]
      .map((h) => `${h} as _${h}`)
      .join(', ')} } from 'vue/vapor';`
  if (helpers.size)
    preamble = `import { ${[...helpers]
      .map((h) => `${h} as _${h}`)
      .join(', ')} } from 'vue';`

  if (!isSetupInlined) {
    ctx.code = preamble + ctx.code
  }

  return {
    code: ctx.code,
    ast: ir as any,
    preamble,
    map: ctx.map ? ctx.map.toJSON() : undefined,
  }
}

function genOperation(oper: OperationNode, context: CodegenContext) {
  const { vaporHelper, push, pushWithNewline } = context
  // TODO: cache old value
  switch (oper.type) {
    case IRNodeTypes.SET_PROP: {
      pushWithNewline(`${vaporHelper('setAttr')}(n${oper.element}, `)
      genExpression(oper.name, context)
      push(`, undefined, `)
      genExpression(oper.value, context)
      push(')')
      return
    }

    case IRNodeTypes.SET_TEXT: {
      pushWithNewline(`${vaporHelper('setText')}(n${oper.element}, undefined, `)
      genExpression(oper.value, context)
      push(')')
      return
    }

    case IRNodeTypes.SET_EVENT: {
      pushWithNewline(`${vaporHelper('on')}(n${oper.element}, `)
      genExpression(oper.name, context)
      push(', ')

      const hasModifiers = oper.modifiers.length
      hasModifiers && push(`${vaporHelper('withModifiers')}(`)
      genExpression(oper.value, context)
      hasModifiers && push(`, ${genArrayExpression(oper.modifiers)})`)

      push(')')
      return
    }

    case IRNodeTypes.SET_HTML: {
      pushWithNewline(`${vaporHelper('setHtml')}(n${oper.element}, undefined, `)
      genExpression(oper.value, context)
      push(')')
      return
    }

    case IRNodeTypes.CREATE_TEXT_NODE: {
      pushWithNewline(`const n${oper.id} = ${vaporHelper('createTextNode')}(`)
      genExpression(oper.value, context)
      push(')')
      return
    }

    case IRNodeTypes.INSERT_NODE: {
      const elements = ([] as number[]).concat(oper.element)
      let element = elements.map((el) => `n${el}`).join(', ')
      if (elements.length > 1) element = `[${element}]`
      pushWithNewline(
        `${vaporHelper('insert')}(${element}, n${
          oper.parent
        }${`, n${oper.anchor}`})`,
      )
      return
    }
    case IRNodeTypes.PREPEND_NODE: {
      pushWithNewline(
        `${vaporHelper('prepend')}(n${oper.parent}, ${oper.elements
          .map((el) => `n${el}`)
          .join(', ')})`,
      )
      return
    }
    case IRNodeTypes.APPEND_NODE: {
      pushWithNewline(
        `${vaporHelper('append')}(n${oper.parent}, ${oper.elements
          .map((el) => `n${el}`)
          .join(', ')})`,
      )
      return
    }
    default:
      return checkNever(oper)
  }
}

function genChildren(children: IRDynamicChildren) {
  let code = ''
  // TODO
  let offset = 0
  for (const [index, child] of Object.entries(children)) {
    const childrenLength = Object.keys(child.children).length
    if (child.ghost && child.placeholder === null && childrenLength === 0) {
      offset--
      continue
    }

    code += ` ${Number(index) + offset}: [`

    const id = child.ghost ? child.placeholder : child.id
    if (id !== null) code += `n${id}`

    const childrenString = childrenLength && genChildren(child.children)
    if (childrenString) code += `, ${childrenString}`

    code += '],'
  }

  if (!code) return ''
  return `{${code}}`
}

// TODO: other types (not only string)
function genArrayExpression(elements: string[]) {
  return `[${elements.map((it) => JSON.stringify(it)).join(', ')}]`
}

function genExpression(
  exp: IRExpression,
  {
    inline,
    prefixIdentifiers,
    bindingMetadata,
    vaporHelper,
    push,
  }: CodegenContext,
) {
  if (isString(exp)) return push(exp)

  // TODO NodeTypes.COMPOUND_EXPRESSION
  if (exp.type === NodeTypes.COMPOUND_EXPRESSION) return

  let { content } = exp
  let name: string | undefined

  if (exp.isStatic) {
    content = JSON.stringify(content)
  } else {
    switch (bindingMetadata[content]) {
      case BindingTypes.SETUP_REF:
        content += '.value'
        break
      case BindingTypes.SETUP_MAYBE_REF:
        content = `${vaporHelper('unref')}(${content})`
        break
    }
    if (prefixIdentifiers && !inline) {
      if (isSimpleIdentifier(content)) name = content
      content = `_ctx.${content}`
    }
  }

  push(content, NewlineType.None, exp.loc, name)
}
