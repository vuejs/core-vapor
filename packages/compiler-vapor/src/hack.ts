import type { Prettify } from '@vue/shared'
import type { DirectiveTransform, NodeTransform } from './transform'

type Overwrite<T, U> = Pick<T, Exclude<keyof T, keyof U>> &
  Pick<U, Extract<keyof U, keyof T>>

export type HackOptions<T> = Prettify<
  Overwrite<
    T,
    {
      nodeTransforms?: NodeTransform[]
      directiveTransforms?: Record<string, DirectiveTransform | undefined>
    }
  >
>
