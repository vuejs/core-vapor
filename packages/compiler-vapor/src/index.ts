export { parse } from '@vue/compiler-dom'
export { transform } from './transform'
export { generate } from './generate'
export { compile, type CompilerOptions } from './compile'
export * from './ir'
export * from './errors'
