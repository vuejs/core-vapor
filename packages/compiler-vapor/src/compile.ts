import {
  type CompilerOptions as BaseCompilerOptions,
  ErrorCodes,
  type RootNode,
  createCompilerError,
  defaultOnError,
  parse,
} from '@vue/compiler-dom'
import { extend, isString } from '@vue/shared'
import {
  type DirectiveTransform,
  type NodeTransform,
  transform,
} from './transform'
import { type VaporCodegenResult, generate } from './generate'
import { transformOnce } from './transforms/vOnce'
import { transformElement } from './transforms/transformElement'
import { transformVHtml } from './transforms/vHtml'
import { transformVText } from './transforms/vText'
import { transformVBind } from './transforms/vBind'
import { transformVOn } from './transforms/vOn'
import { transformVShow } from './transforms/vShow'
import { transformRef } from './transforms/transformRef'
import { transformInterpolation } from './transforms/transformInterpolation'
import type { HackOptions } from './ir'
import { transformVModel } from './transforms/vModel'
import { transformVIf } from './transforms/vIf'
import { transformVFor } from './transforms/vFor'

export type CompilerOptions = HackOptions<BaseCompilerOptions>

// TODO: copied from @vue/compiler-core
// code/AST -> IR -> JS codegen
export function compile(
  source: string | RootNode,
  options: CompilerOptions = {},
): VaporCodegenResult {
  const onError = options.onError || defaultOnError
  const isModuleMode = options.mode === 'module'
  /* istanbul ignore if */
  if (__BROWSER__) {
    if (options.prefixIdentifiers === true) {
      onError(createCompilerError(ErrorCodes.X_PREFIX_ID_NOT_SUPPORTED))
    } else if (isModuleMode) {
      onError(createCompilerError(ErrorCodes.X_MODULE_MODE_NOT_SUPPORTED))
    }
  }

  const prefixIdentifiers =
    !__BROWSER__ && (options.prefixIdentifiers === true || isModuleMode)

  // TODO scope id
  // if (options.scopeId && !isModuleMode) {
  //   onError(createCompilerError(ErrorCodes.X_SCOPE_ID_NOT_SUPPORTED))
  // }

  const resolvedOptions = extend({}, options, {
    prefixIdentifiers,
  })
  const ast = isString(source) ? parse(source, resolvedOptions) : source
  const [nodeTransforms, directiveTransforms] =
    getBaseTransformPreset(prefixIdentifiers)

  if (!__BROWSER__ && options.isTS) {
    const { expressionPlugins } = options
    if (!expressionPlugins || !expressionPlugins.includes('typescript')) {
      resolvedOptions.expressionPlugins = [
        ...(expressionPlugins || []),
        'typescript',
      ]
    }
  }

  const ir = transform(
    ast,
    extend({}, resolvedOptions, {
      nodeTransforms: [
        ...nodeTransforms,
        ...(options.nodeTransforms || []), // user transforms
      ],
      directiveTransforms: extend(
        {},
        directiveTransforms,
        options.directiveTransforms || {}, // user transforms
      ),
    }),
  )

  return generate(ir, resolvedOptions)
}

export type TransformPreset = [
  NodeTransform[],
  Record<string, DirectiveTransform>,
]

export function getBaseTransformPreset(
  prefixIdentifiers?: boolean,
): TransformPreset {
  return [
    [
      transformOnce,
      transformRef,
      transformInterpolation,
      transformVIf,
      transformVFor,
      transformElement,
    ],
    {
      bind: transformVBind,
      on: transformVOn,
      html: transformVHtml,
      text: transformVText,
      show: transformVShow,
      model: transformVModel,
    },
  ]
}
