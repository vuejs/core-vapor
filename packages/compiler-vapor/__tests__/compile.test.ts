import {
  type RootNode,
  BindingTypes,
  ErrorCodes,
  DOMErrorCodes,
} from '@vue/compiler-dom'
import { type CompilerOptions, compile as _compile } from '../src'

function compile(template: string | RootNode, options: CompilerOptions = {}) {
  let { code } = _compile(template, options)
  return code
}

describe('compile', () => {
  test('static template', async () => {
    const code = await compile(
      `<div>
        <p>hello</p>
        <input />
        <span />
      </div>`,
    )
    expect(code).matchSnapshot()
  })

  test('dynamic root', async () => {
    const code = await compile(`{{ 1 }}{{ 2 }}`)
    expect(code).matchSnapshot()
  })

  test('dynamic root nodes and interpolation', async () => {
    const code = await compile(
      `<button @click="handleClick" :id="count">{{count}}foo{{count}}foo{{count}} </button>`,
    )
    expect(code).matchSnapshot()
  })

  test('static + dynamic root', async () => {
    const code = await compile(
      `{{ 1 }}{{ 2 }}3{{ 4 }}{{ 5 }}6{{ 7 }}{{ 8 }}9{{ 'A' }}{{ 'B' }}`,
    )
    expect(code).matchSnapshot()
  })

  test('fragment', async () => {
    const code = await compile(`<p/><span/><div/>`)
    expect(code).matchSnapshot()
  })

  test('bindings', async () => {
    const code = await compile(`<div>count is {{ count }}.</div>`, {
      bindingMetadata: {
        count: BindingTypes.SETUP_REF,
      },
    })
    expect(code).matchSnapshot()
  })

  describe('directives', () => {
    describe('v-bind', () => {
      test('simple expression', async () => {
        const code = await compile(`<div :id="id"></div>`, {
          bindingMetadata: {
            id: BindingTypes.SETUP_REF,
          },
        })
        expect(code).matchSnapshot()
      })

      test('should error if no expression', async () => {
        const onError = vi.fn()
        await compile(`<div v-bind:arg />`, { onError })

        expect(onError.mock.calls[0][0]).toMatchObject({
          code: ErrorCodes.X_V_BIND_NO_EXPRESSION,
          loc: {
            start: {
              line: 1,
              column: 6,
            },
            end: {
              line: 1,
              column: 16,
            },
          },
        })
      })
    })

    describe('v-on', () => {
      test('simple expression', async () => {
        const code = await compile(`<div @click="handleClick"></div>`, {
          bindingMetadata: {
            handleClick: BindingTypes.SETUP_CONST,
          },
        })
        expect(code).matchSnapshot()
      })

      test('should error if no expression AND no modifier', async () => {
        const onError = vi.fn()
        await compile(`<div v-on:click />`, { onError })
        expect(onError.mock.calls[0][0]).toMatchObject({
          code: ErrorCodes.X_V_ON_NO_EXPRESSION,
          loc: {
            start: {
              line: 1,
              column: 6,
            },
            end: {
              line: 1,
              column: 16,
            },
          },
        })
      })

      test('event modifier', async () => {
        const code = await compile(
          `<a @click.stop="handleEvent"></a>
            <form @submit.prevent="handleEvent"></form>
            <a @click.stop.prevent="handleEvent"></a>
            <div @click.self="handleEvent"></div> 
            <div @click.capture="handleEvent"></div> 
            <a @click.once="handleEvent"></a>
            <div @scroll.passive="handleEvent"></div>
            <input @click.right="handleEvent" />
            <input @click.left="handleEvent" />
            <input @click.middle="handleEvent" />
            <input @click.enter.right="handleEvent" />
            <input @keyup.enter="handleEvent" />
            <input @keyup.tab="handleEvent" />
            <input @keyup.delete="handleEvent" />
            <input @keyup.esc="handleEvent" />
            <input @keyup.space="handleEvent" />
            <input @keyup.up="handleEvent" />
            <input @keyup.down="handleEvent" />
            <input @keyup.left="handleEvent" />
            <input @keyup.middle="submit" />
            <input @keyup.middle.self="submit" />
            <input @keyup.self.enter="handleEvent" />`,
          {
            bindingMetadata: {
              handleEvent: BindingTypes.SETUP_CONST,
            },
          },
        )
        expect(code).matchSnapshot()
      })
    })

    describe('v-html', () => {
      test('simple expression', async () => {
        const code = await compile(`<div v-html="code"></div>`, {
          bindingMetadata: {
            code: BindingTypes.SETUP_REF,
          },
        })
        expect(code).matchSnapshot()
      })

      test('should raise error and ignore children when v-html is present', async () => {
        const onError = vi.fn()
        const code = await compile(`<div v-html="test">hello</div>`, {
          onError,
        })
        expect(code).matchSnapshot()
        expect(onError.mock.calls).toMatchObject([
          [{ code: DOMErrorCodes.X_V_HTML_WITH_CHILDREN }],
        ])
      })

      test('should raise error if has no expression', async () => {
        const onError = vi.fn()
        const code = await compile(`<div v-html></div>`, {
          onError,
        })
        expect(code).matchSnapshot()
        expect(onError.mock.calls).toMatchObject([
          [{ code: DOMErrorCodes.X_V_HTML_NO_EXPRESSION }],
        ])
      })
    })

    describe('v-text', () => {
      test('simple expression', async () => {
        const code = await compile(`<div v-text="str"></div>`, {
          bindingMetadata: {
            str: BindingTypes.SETUP_REF,
          },
        })
        expect(code).matchSnapshot()
      })

      test('no expression', async () => {
        const onError = vi.fn()
        const code = await compile(`<div v-text></div>`, { onError })
        expect(code).matchSnapshot()
        expect(onError.mock.calls).toMatchObject([
          [{ code: DOMErrorCodes.X_V_TEXT_NO_EXPRESSION }],
        ])
      })
    })

    describe('v-once', () => {
      test('basic', async () => {
        const code = await compile(
          `<div v-once>
            {{ msg }}
            <span :class="clz" />
          </div>`,
          {
            bindingMetadata: {
              msg: BindingTypes.SETUP_REF,
              clz: BindingTypes.SETUP_REF,
            },
          },
        )
        expect(code).matchSnapshot()
      })

      test('as root node', async () => {
        const code = await compile(`<div :id="foo" v-once />`)
        expect(code).toMatchSnapshot()
        expect(code).not.contains('effect')
      })
    })

    describe('v-pre', () => {
      test('basic', async () => {
        const code = await compile(
          `<div v-pre :id="foo"><Comp/>{{ bar }}</div>\n`,
          {
            bindingMetadata: {
              foo: BindingTypes.SETUP_REF,
              bar: BindingTypes.SETUP_REF,
            },
          },
        )

        expect(code).toMatchSnapshot()
        expect(code).contains(
          JSON.stringify('<div :id="foo"><Comp></Comp>{{ bar }}</div>'),
        )
        expect(code).not.contains('effect')
      })

      // TODO: support multiple root nodes and components
      test('should not affect siblings after it', async () => {
        const code = await compile(
          `<div v-pre :id="foo"><Comp/>{{ bar }}</div>\n` +
            `<div :id="foo"><Comp/>{{ bar }}</div>`,
          {
            bindingMetadata: {
              foo: BindingTypes.SETUP_REF,
              bar: BindingTypes.SETUP_REF,
            },
          },
        )

        expect(code).toMatchSnapshot()
        // Waiting for TODO, There should be more here.
      })

      // TODO: support multiple root nodes and components
      test('self-closing v-pre', async () => {
        const code = await compile(
          `<div v-pre/>\n<div :id="foo"><Comp/>{{ bar }}</div>`,
        )

        expect(code).toMatchSnapshot()
        expect(code).contains('<div></div><div><Comp></Comp></div>')
        // Waiting for TODO, There should be more here.
      })
    })
  })
})
