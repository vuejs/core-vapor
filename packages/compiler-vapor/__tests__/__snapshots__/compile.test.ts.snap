// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`compile > bindings 1`] = `
"import { template as _template, children as _children, createTextNode as _createTextNode, insert as _insert, effect as _effect, setText as _setText } from 'vue/vapor';

export function render(_ctx) {
  const t0 = _template(\\"<div>count is <!>.</div>\\")
  const n0 = t0()
  const { 0: [n3, { 1: [n2],}],} = _children(n0)
  const n1 = _createTextNode(count.value)
  _insert(n1, n3, n2)
  _effect(() => {
    _setText(n1, undefined, count.value)
  })
  return n0
}"
`;

exports[`compile > directives > v-bind > simple expression 1`] = `
"import { template as _template, children as _children, effect as _effect, setAttr as _setAttr } from 'vue/vapor';

export function render(_ctx) {
  const t0 = _template(\\"<div></div>\\")
  const n0 = t0()
  const { 0: [n1],} = _children(n0)
  _effect(() => {
    _setAttr(n1, \\"id\\", undefined, id.value)
  })
  return n0
}"
`;

exports[`compile > directives > v-html > should raise error and ignore children when v-html is present 1`] = `
"import { template as _template, children as _children, effect as _effect, setHtml as _setHtml } from 'vue/vapor';

export function render(_ctx) {
  const t0 = _template(\\"<div></div>\\")
  const n0 = t0()
  const { 0: [n1],} = _children(n0)
  _effect(() => {
    _setHtml(n1, undefined, test)
  })
  return n0
}"
`;

exports[`compile > directives > v-html > should raise error if has no expression 1`] = `
"import { template as _template, children as _children, setHtml as _setHtml } from 'vue/vapor';

export function render(_ctx) {
  const t0 = _template(\\"<div></div>\\")
  const n0 = t0()
  const { 0: [n1],} = _children(n0)
  _setHtml(n1, undefined, \\"\\")
  return n0
}"
`;

exports[`compile > directives > v-html > simple expression 1`] = `
"import { template as _template, children as _children, effect as _effect, setHtml as _setHtml } from 'vue/vapor';

export function render(_ctx) {
  const t0 = _template(\\"<div></div>\\")
  const n0 = t0()
  const { 0: [n1],} = _children(n0)
  _effect(() => {
    _setHtml(n1, undefined, code.value)
  })
  return n0
}"
`;

exports[`compile > directives > v-on > event modifier 1`] = `
"import { template as _template, children as _children, effect as _effect, on as _on, withModifiers as _withModifiers } from 'vue/vapor';

export function render(_ctx) {
  const t0 = _template(\\"<div></div>\\")
  const n0 = t0()
  const { 0: [n1],} = _children(n0)
  _effect(() => {
    _on(n1, \\"click\\", _withModifiers(handleClick, [\\"prevent\\", \\"stop\\"]))
  })
  return n0
}"
`;

exports[`compile > directives > v-on > simple expression 1`] = `
"import { template as _template, children as _children, effect as _effect, on as _on } from 'vue/vapor';

export function render(_ctx) {
  const t0 = _template(\\"<div></div>\\")
  const n0 = t0()
  const { 0: [n1],} = _children(n0)
  _effect(() => {
    _on(n1, \\"click\\", handleClick)
  })
  return n0
}"
`;

exports[`compile > directives > v-once > as root node 1`] = `
"import { template as _template, children as _children, setAttr as _setAttr } from 'vue/vapor';

export function render(_ctx) {
  const t0 = _template(\\"<div></div>\\")
  const n0 = t0()
  const { 0: [n1],} = _children(n0)
  _setAttr(n1, \\"id\\", undefined, foo)
  return n0
}"
`;

exports[`compile > directives > v-once > basic 1`] = `
"import { template as _template, children as _children, createTextNode as _createTextNode, setText as _setText, setAttr as _setAttr, prepend as _prepend } from 'vue/vapor';

export function render(_ctx) {
  const t0 = _template(\\"<div> <span></span></div>\\")
  const n0 = t0()
  const { 0: [n3, { 1: [n2],}],} = _children(n0)
  const n1 = _createTextNode(msg.value)
  _setText(n1, undefined, msg.value)
  _setAttr(n2, \\"class\\", undefined, clz.value)
  _prepend(n3, n1)
  return n0
}"
`;

exports[`compile > directives > v-pre > basic 1`] = `
"import { template as _template } from 'vue/vapor';

export function render(_ctx) {
  const t0 = _template(\\"<div :id=\\\\\\"foo\\\\\\"><Comp></Comp>{{ bar }}</div>\\")
  const n0 = t0()
  return n0
}"
`;

exports[`compile > directives > v-pre > self-closing v-pre 1`] = `
"import { template as _template, children as _children, createTextNode as _createTextNode, append as _append, effect as _effect, setAttr as _setAttr, setText as _setText } from 'vue/vapor';

export function render(_ctx) {
  const t0 = _template(\\"<div></div><div><Comp></Comp></div>\\")
  const n0 = t0()
  const { 1: [n1],} = _children(n0)
  const n2 = _createTextNode(bar)
  _append(n1, n2)
  _effect(() => {
    _setAttr(n1, \\"id\\", undefined, foo)
  })
  _effect(() => {
    _setText(n2, undefined, bar)
  })
  return n0
}"
`;

exports[`compile > directives > v-pre > should not affect siblings after it 1`] = `
"import { template as _template, children as _children, createTextNode as _createTextNode, append as _append, effect as _effect, setAttr as _setAttr, setText as _setText } from 'vue/vapor';

export function render(_ctx) {
  const t0 = _template(\\"<div :id=\\\\\\"foo\\\\\\"><Comp></Comp>{{ bar }}</div><div><Comp></Comp></div>\\")
  const n0 = t0()
  const { 1: [n1],} = _children(n0)
  const n2 = _createTextNode(bar.value)
  _append(n1, n2)
  _effect(() => {
    _setAttr(n1, \\"id\\", undefined, foo.value)
  })
  _effect(() => {
    _setText(n2, undefined, bar.value)
  })
  return n0
}"
`;

exports[`compile > directives > v-text > no expression 1`] = `
"import { template as _template, children as _children, setText as _setText } from 'vue/vapor';

export function render(_ctx) {
  const t0 = _template(\\"<div></div>\\")
  const n0 = t0()
  const { 0: [n1],} = _children(n0)
  _setText(n1, undefined, \\"\\")
  return n0
}"
`;

exports[`compile > directives > v-text > simple expression 1`] = `
"import { template as _template, children as _children, effect as _effect, setText as _setText } from 'vue/vapor';

export function render(_ctx) {
  const t0 = _template(\\"<div></div>\\")
  const n0 = t0()
  const { 0: [n1],} = _children(n0)
  _effect(() => {
    _setText(n1, undefined, str.value)
  })
  return n0
}"
`;

exports[`compile > dynamic root 1`] = `
"import { fragment as _fragment, createTextNode as _createTextNode, append as _append, effect as _effect, setText as _setText } from 'vue/vapor';

export function render(_ctx) {
  const t0 = _fragment()

  const n0 = t0()
  const n1 = _createTextNode(1)
  const n2 = _createTextNode(2)
  _append(n0, n1, n2)
  _effect(() => {
    _setText(n1, undefined, 1)
  })
  _effect(() => {
    _setText(n2, undefined, 2)
  })
  return n0
}"
`;

exports[`compile > dynamic root nodes and interpolation 1`] = `
"import { template as _template, children as _children, createTextNode as _createTextNode, prepend as _prepend, insert as _insert, append as _append, effect as _effect, on as _on, setAttr as _setAttr, setText as _setText } from 'vue/vapor';

export function render(_ctx) {
  const t0 = _template(\\"<button>foo<!>foo</button>\\")
  const n0 = t0()
  const { 0: [n1, { 1: [n5],}],} = _children(n0)
  const n2 = _createTextNode(count)
  const n3 = _createTextNode(count)
  const n4 = _createTextNode(count)
  _prepend(n1, n2)
  _insert(n3, n1, n5)
  _append(n1, n4)
  _effect(() => {
    _on(n1, \\"click\\", handleClick)
  })
  _effect(() => {
    _setAttr(n1, \\"id\\", undefined, count)
  })
  _effect(() => {
    _setText(n2, undefined, count)
  })
  _effect(() => {
    _setText(n3, undefined, count)
  })
  _effect(() => {
    _setText(n4, undefined, count)
  })
  return n0
}"
`;

exports[`compile > fragment 1`] = `
"import { template as _template } from 'vue/vapor';

export function render(_ctx) {
  const t0 = _template(\\"<p></p><span></span><div></div>\\")
  const n0 = t0()
  return n0
}"
`;

exports[`compile > static + dynamic root 1`] = `
"import { template as _template, children as _children, createTextNode as _createTextNode, prepend as _prepend, insert as _insert, append as _append, effect as _effect, setText as _setText } from 'vue/vapor';

export function render(_ctx) {
  const t0 = _template(\\"3<!>6<!>9\\")
  const n0 = t0()
  const { 1: [n9], 3: [n10],} = _children(n0)
  const n1 = _createTextNode(1)
  const n2 = _createTextNode(2)
  const n3 = _createTextNode(4)
  const n4 = _createTextNode(5)
  const n5 = _createTextNode(7)
  const n6 = _createTextNode(8)
  const n7 = _createTextNode('A')
  const n8 = _createTextNode('B')
  _prepend(n0, n1, n2)
  _insert([n3, n4], n0, n9)
  _insert([n5, n6], n0, n10)
  _append(n0, n7, n8)
  _effect(() => {
    _setText(n1, undefined, 1)
  })
  _effect(() => {
    _setText(n2, undefined, 2)
  })
  _effect(() => {
    _setText(n3, undefined, 4)
  })
  _effect(() => {
    _setText(n4, undefined, 5)
  })
  _effect(() => {
    _setText(n5, undefined, 7)
  })
  _effect(() => {
    _setText(n6, undefined, 8)
  })
  _effect(() => {
    _setText(n7, undefined, 'A')
  })
  _effect(() => {
    _setText(n8, undefined, 'B')
  })
  return n0
}"
`;

exports[`compile > static template 1`] = `
"import { template as _template } from 'vue/vapor';

export function render(_ctx) {
  const t0 = _template(\\"<div><p>hello</p><input><span></span></div>\\")
  const n0 = t0()
  return n0
}"
`;
